pkg/storage/oss/oss.go

```diff
@@ -0,0 +1,673 @@
+	imageRegistryAccessKeySecret  = "access_key_secret"
+
+	envRegistryStorage                   = "REGISTRY_STORAGE"
+	envRegistryStorageOssBucket          = "REGISTRY_STORAGE_OSS_BUCKET"
+	envRegistryStorageOssRegion          = "REGISTRY_STORAGE_OSS_REGION"
+	envRegistryStorageOssEndpoint        = "REGISTRY_STORAGE_OSS_ENDPOINT"
+	envRegistryStorageOssEncrypt         = "REGISTRY_STORAGE_OSS_ENCRYPT"
+	envRegistryStorageOssInternal        = "REGISTRY_STORAGE_OSS_INTERNAL"
+	envRegistryStorageOssAccessKeyId     = "REGISTRY_STORAGE_OSS_ACCESSKEYID"
+	envRegistryStorageOssAccessKeySecret = "REGISTRY_STORAGE_OSS_ACCESSKEYSECRET"
+)
+
+var mutex sync.Mutex
+
+type driver struct {
+	Context     context.Context
+	Config      *imageregistryv1.ImageRegistryConfigStorageAlibabaOSS
+	Listers     *regopclient.Listers
+	credentials *credentials.AccessKeyCredential
+
+	// roundTripper is used only during tests.
+	roundTripper http.RoundTripper
+}
+
+// NewDriver creates a new OSS storage driver
+// Used during bootstrapping
+func NewDriver(ctx context.Context, c *imageregistryv1.ImageRegistryConfigStorageAlibabaOSS, listers *regopclient.Listers) *driver {
+	return &driver{
+		Context: ctx,
+		Config:  c,
+		Listers: listers,
+	}
+}
+
+// UpdateEffectiveConfig updates the driver's local effective OSS configuration
+// based on infrastructure settings and any custom overrides.
+func (d *driver) UpdateEffectiveConfig() error {
+	effectiveConfig := d.Config.DeepCopy()
+
+	if effectiveConfig == nil {
+		effectiveConfig = &imageregistryv1.ImageRegistryConfigStorageAlibabaOSS{}
+	}
+
+	// Load infrastructure values
+	infra, err := util.GetInfrastructure(d.Listers)
+	if err != nil {
+		return err
+	}
+
+	var clusterRegion string
+	if infra.Status.PlatformStatus != nil && infra.Status.PlatformStatus.Type == configv1.AlibabaCloudPlatformType {
+		clusterRegion = infra.Status.PlatformStatus.AlibabaCloud.Region
+	}
+
+	// Use cluster defaults when custom config doesn't define values
+	if d.Config == nil || len(effectiveConfig.Region) == 0 {
+		effectiveConfig.Region = clusterRegion
+	}
+
+	err = d.getCredentialsConfigData()
+	if err != nil {
+		return err
+	}
+
+	d.Config = effectiveConfig.DeepCopy()
+
+	return nil
+}
+
+func fetchCredentialsIniFromSecret(secret *corev1.Secret) (auth.Credential, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	creds, ok := secret.Data[imageRegistrySecretDataKey]
+	if !ok {
+		return nil, fmt.Errorf("failed to fetch key 'credentials' in secret data")
+	}
+
+	f, err := ioutil.TempFile("", "alibaba-creds-*")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	defer os.Remove(f.Name())
+	defer os.Unsetenv(provider.ENVCredentialFile)
+	_, err = f.Write(creds)
+	if err != nil {
+		return nil, err
+	}
+	os.Setenv(provider.ENVCredentialFile, f.Name())
+	// use Alibaba provider initialization
+	p := provider.NewProfileProvider("default")
+	// return a valid auth credential
+	return p.Resolve()
```

Comment:
Is there a way to get `auth.Credential` without creating a temporary file?

Reply:
@kwoodson  please help explain this. I think we get credentials from cco.

Reply:
Currently the INI file parsing that is provided through Alibaba's API is environment variable driven. 
https://github.com/aliyun/alibaba-cloud-sdk-go/blob/master/sdk/auth/credentials/provider/profile_credentials.go#L41

We could parse the INI ourselves but using the library allows us to use this method across the openshift components.